examples/hashmap.dx

```diff
@@ -0,0 +1,266 @@
+
+
+-- Can't require type-classes in type definitions?
+-- data ListDict : (_:Eq k) ?=> k:Type v:Type =
+data ListDict k:Type v:Type =
+  AsListDict vals:(List (k&v))
+
+eg_listdict = AsListDict $ AsList _ [('a', 10), ('b', 20), ('c', 30)]
+
+:p eg_listdict
+
+def push (collectionL:List a) (datum:a) : List a =
+    (AsList n collection) = collectionL
+    neo_n = Fin (n+1)
+    AsList _  $ for i:neo_n. case (ordinal i < n) of
+        True -> collection.(ordinal i@(Fin n))
+        False -> datum
+
+:p push (AsList _ "abc") 'x'
+
+def tryGetListDict (_:Eq k) ?=> (dict:ListDict k v) (key:k) : Maybe v =
+    (AsListDict (AsList _ vals)) = dict
+    fold Nothing \i state. 
+        case ((fst vals.i) == key) of
+            False -> state
+            True -> Just $ snd vals.i
+
+:p tryGetListDict eg_listdict 'a'
+
+:p tryGetListDict eg_listdict 'x'
+
+'store dict key val
+Stores the value in the dict at the key.
+Note that we never actually delete anything, and it grows without bounds
+This could be improved.
+
+def storeListDict (dict:ListDict k v) (key:k) (val:v) : ListDict k v =
+    (AsListDict list) = dict
+    pair = (key, val)
+    AsListDict (push list pair)
+
+
+eg_listdict2 = storeListDict eg_listdict 'x' 40
+
+:p eg_listdict2
+
+
+:p tryGetListDict eg_listdict2 'x'
+
+
+def emptyListDict (k:Type) (v:Type) : ListDict k v =
+    empty_table : Fin 0 => (k&v) = []
+    AsListDict $ AsList _ empty_table
+
+:p emptyListDict Int Char
+
+:p storeListDict (emptyListDict Int Char) 100 'a'
+
+' ## Hash
+We need a hash function for everything
+
+
+data Hashable a:Type = MkHashable (Eq a) (a->Int->Int) -- equals, hash
+
+@superclass
+def eqFromHashable (d:Hashable a) : Eq a = case d of MkHashable eqDict _ -> eqDict
+
+def ahash (d:Hashable a) ?=> (x:a) (seed:Int) : Int = case d of MkHashable _ hsh  -> hsh x seed
+
+thehash: Hashable a ?=> a->Int = \x. ahash x 0
+
+-- composition operator 1 arg and 2 arg forms (would be `of` but that is a reserved word)
+ov : (a->b)->(c->a)->(c->b) = \f g x. f(g x)
+ov2: (c->d)->(a->b->c)->(a->b->d) = \f g x y. f(g x y)
+
+@instance int32Hashable : Hashable Int32 = (MkHashable
+    int32Eq
+    (I64ToI `ov2` hash `ov` IToI64)
+)
+
+:p thehash 1
+
+@instance word8Hashable : Hashable Word8 = (MkHashable
+    word8Eq
+    (ahash `ov` W8ToI)
+)
+
+:p thehash 'a'
+
+
+@instance int64Hashable : Hashable Int64 = (MkHashable
+    int64Eq
+    (ahash `ov` I64ToI)
+)
+
+:p thehash $ IToI64 213
+
+@instance boolHashable : Hashable Bool = (MkHashable
+    boolEq
+    (ahash `ov` BToI)
+)
+:p thehash True
+:p thehash False
+
+
+@instance unitHashable : Hashable Unit = (MkHashable
+    unitEq
+    (\x seed. seed)
+)
+
+:p thehash ()
+
+' due to witchcraft defining a type-class on s 2-tuple defines it on all tuples
+
+@instance pairHashable : Hashable a ?=> Hashable b ?=>  Hashable (a & b) = (MkHashable
+    pairEq
+    (\(a,b). ahash a `ov` ahash b)
```

Comment:
I feel like this should be doable with the right might of `uncurry`, `ov`, `ov2` and maybe `flip`.
But probably much harder to read.